Appliquer une Analyse en composantes principales sur Iris Dataset ou wine dataset.
Les étapes sont présentes dans le cours. N’oubliez pas de plotter les histogram de
chaque variable du jeu de données utilisé

In [26]:
import numpy as  np
from sklearn.datasets import load_iris
import pandas as pd


data = load_iris()
data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [27]:
iris_data = data.data.copy()
moy = iris_data.mean(axis=0)


In [29]:
iris_centered = (iris_data - moy)/iris_data.std(axis=0)
pd.DataFrame(np.cov(iris_centered, rowvar=False, bias=True))

,0,1,2,3
0,1.000000,-0.117570,0.871754,0.817941
1,-0.117570,1.000000,-0.428440,-0.366126
2,0.871754,-0.428440,1.000000,0.962865
3,0.817941,-0.366126,0.962865,1.000000


In [30]:
cov_matrix = np.cov(iris_centered, rowvar=False, bias=True)

# calculer les valeurs et les vecteurs propres de la matrice de la covariance



In [ ]:
eig_values, eig_vectors = np.linalg.eig(cov_matrix)
eig_values


array([0.72962445, 0.22850762, 0.03668922, 0.00517871])

In [37]:
var_expli = eig_values / np.sum(eig_values)
var_expli

array([0.72962445, 0.22850762, 0.03668922, 0.00517871])

In [38]:
var_expli[:2].sum()

np.float64(0.9581320720000164)

In [44]:
eig_vectors

array([[ 0.52106591, -0.37741762, -0.71956635,  0.26128628],
       [-0.26934744, -0.92329566,  0.24438178, -0.12350962],
       [ 0.5804131 , -0.02449161,  0.14212637, -0.80144925],
       [ 0.56485654, -0.06694199,  0.63427274,  0.52359713]])

In [52]:
# on recupere les deux premieres colonnes
P =  np.vstack((eig_vectors[:,0], eig_vectors[:,1])) 

In [ ]:
# on recupere les deux premieres lignes
np.vstack((eig_vectors[0], eig_vectors[1])) 

In [53]:
Z = iris_centered @ P.T
Z

array([[-2.26470281, -0.4800266 ],
       [-2.08096115,  0.67413356],
       [-2.36422905,  0.34190802],
       [-2.29938422,  0.59739451],
       [-2.38984217, -0.64683538],
       [-2.07563095, -1.48917752],
       [-2.44402884, -0.0476442 ],
       [-2.23284716, -0.22314807],
       [-2.33464048,  1.11532768],
       [-2.18432817,  0.46901356],
       [-2.1663101 , -1.04369065],
       [-2.32613087, -0.13307834],
       [-2.2184509 ,  0.72867617],
       [-2.6331007 ,  0.96150673],
       [-2.1987406 , -1.86005711],
       [-2.26221453, -2.68628449],
       [-2.2075877 , -1.48360936],
       [-2.19034951, -0.48883832],
       [-1.898572  , -1.40501879],
       [-2.34336905, -1.12784938],
       [-1.914323  , -0.40885571],
       [-2.20701284, -0.92412143],
       [-2.7743447 , -0.45834367],
       [-1.81866953, -0.08555853],
       [-2.22716331, -0.13725446],
       [-1.95184633,  0.62561859],
       [-2.05115137, -0.24216355],
       [-2.16857717, -0.52714953],
       [-2.13956345,

axe1 = 0.52 sl - 0.26sw + 0.58pl + 0.56 pq

axe2 = -0.37 sl -0.92sw -0.02pl -0.06pq